tools/bpf2c/bpf_code_generator.cpp

```diff
@@ -0,0 +1,641 @@
+            }
+            auto& format = _predicate_format_string[inst.opcode >> 4];
+            if (inst.opcode == EBPF_OP_JA) {
+                std::string target = program_output[i + inst.offset + 1].label;
+                output.lines.push_back(std::string("goto ") + target + std::string(";"));
+            } else if (inst.opcode == EBPF_OP_CALL) {
+                std::string function_name;
+                if (output.relocation.empty()) {
+                    function_name = format_string(
+                        "%s_helpers[%s]",
+                        desired_section,
+                        std::to_string(
+                            functions[std::string("helper_id_") + std::to_string(output.instruction.imm)].index));
+                } else {
+                    function_name = format_string(
+                        "%s_helpers[%s]", desired_section, std::to_string(functions[output.relocation].index));
+                }
+                output.lines.push_back(
+                    get_register_name(0) + std::string(" = ") + function_name + std::string(".address"));
+                output.lines.push_back(std::string("(") + get_register_name(1) + std::string(", "));
+                output.lines.push_back(get_register_name(2) + std::string(", "));
+                output.lines.push_back(get_register_name(3) + std::string(", "));
+                output.lines.push_back(get_register_name(4) + std::string(", "));
+                output.lines.push_back(get_register_name(5) + std::string(");"));
+                output.lines.push_back(
+                    format_string("\tif ((%s.tail_call) && (%s == 0)) return 0;", function_name, get_register_name(0)));
+            } else if (inst.opcode == EBPF_OP_EXIT) {
+                output.lines.push_back(std::string("return ") + get_register_name(0) + std::string(";"));
+            } else {
+                std::string target = program_output[i + inst.offset + 1].label;
+                if (target.empty()) {
+                    throw std::runtime_error("invalid jump target");
+                }
+                std::string predicate = format_string(format, destination, source);
+                output.lines.push_back(format_string("if (%s) goto %s;", predicate, target));
+            }
+        } break;
+        default:
+            throw std::runtime_error("invalid operand");
+        }
+    }
+}
+
+void
+bpf_code_generator::emit_c_code(std::ostream& output_stream)
+{
+    // Emit C file
+    output_stream << "// Do not alter this generated file." << std::endl;
+    output_stream << "// This file was generated from " << path.c_str() << std::endl << std::endl;
+    output_stream << "#include \"bpf2c.h\"" << std::endl << std::endl;
+
+    // Emit import tables
+    if (map_definitions.size() > 0) {
+        output_stream << "static map_entry_t " << desired_section.c_str() << "_maps[] = {" << std::endl;
```

Comment:
Need to research what M4 is.

Reply:
https://en.wikipedia.org/wiki/M4_(computer_language)
but any popular template processor would be ok instead of creating a new one.  Not a blocker for this PR though, and it's just a "consider".